README.md
![repo-size](https://img.shields.io/github/repo-size/vel-san/killing-floor-docker?label=Repo-Size&style=flat-square) [![contrib](https://img.shields.io/github/contributors/vel-san/killing-floor-docker?label=Contributors&style=flat-square)](https://github.com/Vel-San/killing-floor-docker/graphs/contributors) [![release](https://img.shields.io/github/v/release/vel-san/killing-floor-docker?label=Release&style=flat-square)](https://github.com/Vel-San/killing-floor-docker/releases) [![Docker](https://img.shields.io/badge/Docker%20Hub-White--Listed%20Vanilla-Blue?style=flat-square)](https://hub.docker.com/r/vel7an/kf1-docker)

[![CI: Build & Deploy to DockerHub](https://github.com/Vel-San/killing-floor-docker/actions/workflows/build_deploy.yml/badge.svg?branch=master)](https://github.com/Vel-San/killing-floor-docker/actions/workflows/build_deploy.yml) [![CI: Container Scanner](https://github.com/Vel-San/killing-floor-docker/actions/workflows/trivy-analysis.yml/badge.svg)](https://github.com/Vel-San/killing-floor-docker/actions/workflows/trivy-analysis.yml)

[![](https://cache.gametracker.com/server_info/nn.h4ck.me:7707/b_560_95_1.png)](https://www.gametracker.com/server_info/nn.h4ck.me:7707/)

- [Killing Floor 1: Dockerized!](#killing-floor-1-dockerized)
  - [Important Notes](#important-notes)
    - [OSX 💻 Users](#osx--users)
    - [Multi-Server 🌩 support](#multi-server--support)
  - [Docker Run Variables](#docker-run-variables)
  - [Build Command](#build-command)
  - [Run Command](#run-command)
  - [Useful docker commads](#useful-docker-commads)

# Killing Floor 1: Dockerized!

## Important Notes

### OSX 💻 Users 

If, for any reason, you are encountering "Operation not permitted" issues while running, make sure you do the following:

1. In your Docker settings, disable/uncheck `Use gRPC FUSE for file sharing` and hit Restart to Apply changes
2. If the above doesn't solve your issue, go to your Machine Settings > Security & Privacy > Give `Full Disk Usage` to Docker.app (If this still doesn't work, give your terminal as well)

This should fix this issue 👍

____

### Multi-Server 🌩 support

In case you want to run more than 1 server (e.g. several containers), do not forget to change the exposed ports in the docker run command to match the following:

```
KF_GAME_PORT=
KF_QUERY_PORT=
KF_WEBADMIN_PORT=
```

Also, keep in mind to update the folder you are using for the mounted volume on your local

>-v (your server folder):/home/steam/servers/kf

## Docker Run Variables

When running a container for the first time, you will need to supply a steam login to install the server

- If you get password errors, just put your password in hyphons e.g 'Password_here'
- Make sure the code you provide (if you have 2FA enabled) is newly generated

```bash
STEAM_USER=
STEAM_PASSWORD=
STEAM_CODE=
```

Use these vars to tweak the server settings when a container starts

>NOTE: Removing variables or not putting any values will automatically use the default values, as found [HERE](https://github.com/Vel-San/killing-floor-docker/blob/1ac09a4d55e7369442bb75831d70413d36e9ced9/scripts/kf1_functions.sh#L24)

```bash
# Map Name
KF_MAP=
# 1 Begginer, 2 Normal, 4 Hard, 5 Suicidal, 7 Hell On Earth
KF_DIFFICULTY=
# 0 --> 4 waves, 1 --> 7 waves, 2 --> 10 waves
KF_GAME_LENGTH=
# Server Name
KF_SERVER_NAME=
# Game Password, Do not provide the parameter if you want no password
KF_GAME_PASS=
# Admin Name
KF_ADMIN_NAME=
# WebAdmin Password
KF_ADMIN_PASS=
# Admin E-Mail
KF_ADMIN_EMAIL=
# Message Of The Day - Shows when someone joins the server
KF_MOTD=
# Redirect URL, if not set, defaults to Skillzserver (Includes almost all KF1 Mods)
KF_REDIRECT=
# Number of Max Spectators that can be in your server
KF_SPECTATORS=
# Change ZED skins based on the special events
# Select 1 of the following:
# ET_None -- Default
# ET_SummerSideshow -- for the summer ZEDs
# ET_HillbillyHorror -- for the Halloween ZEDs
# ET_TwistedChristmas -- for the Christmas ZEDs
KF_SPECIMEN_EVENT_TYPE=
# List of your mutators; seperated by commas and NO spaces -- It is recommended to use MutLoader to load your muts
KF_MUTATORS=
# DO. NOT. TOUCH! Port variables if you don't know what you're doing!
# Game Port
KF_GAME_PORT=
# Query Port
KF_QUERY_PORT=
# WebAdmin Port
KF_WEBADMIN_PORT=
```

## Build Command

```bash
docker build -t kf1-docker .
```

## Run Command

- Replace the values in Parenthesis with your own values

Config directly as env vars:

```bash
docker run -d --name kf1 \
        -p 0.0.0.0:7707:7707/udp \
        -p 0.0.0.0:7708:7708/udp \
        -p 0.0.0.0:7717:7717/udp \
        -p 0.0.0.0:28852:28852/udp \
        -p 0.0.0.0:28852:28852/tcp \
        -p 0.0.0.0:8075:8075/tcp \
        -p 0.0.0.0:20560:20560/udp \
        -p 0.0.0.0:20560:20560/tcp \
        -v (your server folder):/home/steam/servers/kf \
        -e STEAM_USER=(user) \
        -e STEAM_PASSWORD=(pass) \
        -e STEAM_CODE=(code) \
        -e KF_MAP=KF-BioticsLab \
        -e KF_DIFFICULTY=4 \
        -e KF_GAME_LENGTH=2 \
        -e KF_GAME_PASS=(pass) \
        -e KF_SERVER_NAME='Killing Floor Server' \
        -e KF_ADMIN_NAME=(name) \
        -e KF_ADMIN_PASS=(pass) \
        -e KF_ADMIN_EMAIL=(email) \
        -e KF_MOTD=Welcome \
        -e KF_REDIRECT=(redirect) \
        -e KF_SPECTATORS=6 \
        -e KF_SPECIMEN_EVENT_TYPE=ET_None \
        -e KF_MUTATORS=(mutators) \
        -e KF_GAME_PORT=7707 \
        -e KF_QUERY_PORT=7717 \
        -e KF_WEBADMIN_PORT=8075
        kf1-docker
```

<details>
  <summary>Copy/Paste (CLICK ME)</summary>

```bash
docker run -d --name kf1 -p 0.0.0.0:7707:7707/udp -p 0.0.0.0:7708:7708/udp -p 0.0.0.0:7717:7717/udp -p 0.0.0.0:28852:28852/udp -p 0.0.0.0:28852:28852/tcp -p 0.0.0.0:8075:8075/tcp -p 0.0.0.0:20560:20560/udp -p 0.0.0.0:20560:20560/tcp -v (your server folder):/home/steam/servers/kf -e STEAM_USER=(user) -e STEAM_PASSWORD=(pass) -e STEAM_CODE=(code) -e KF_MAP=KF-BioticsLab -e KF_DIFFICULTY=4 -e KF_GAME_LENGTH=2 -e KF_GAME_PASS=(pass) -e KF_SERVER_NAME='Killing Floor Server' -e KF_ADMIN_NAME=(name) -e KF_ADMIN_PASS=(pass) -e KF_ADMIN_EMAIL=(email) -e KF_MOTD=Welcome -e KF_REDIRECT=(redirect) -e KF_SPECTATORS=6 -e KF_SPECIMEN_EVENT_TYPE=ET_None -e KF_MUTATORS=(mutators) -e KF_GAME_PORT=7707 -e KF_QUERY_PORT=7717 -e KF_WEBADMIN_PORT=8075 kf1-docker
```

</details>

or config taken from Env Var (see example env_file)

```bash
docker run -d --name kf1 \
        -p 0.0.0.0:7707:7707/udp \
        -p 0.0.0.0:7708:7708/udp \
        -p 0.0.0.0:7717:7717/udp \
        -p 0.0.0.0:28852:28852/udp \
        -p 0.0.0.0:28852:28852/tcp \
        -p 0.0.0.0:8075:8075/tcp \
        -p 0.0.0.0:20560:20560/udp \
        -p 0.0.0.0:20560:20560/tcp \
        -v (your server folder):/home/steam/servers/kf \
        -e STEAM_USER=(user) \
        -e STEAM_PASSWORD=(pass) \
        -e STEAM_CODE=(code) \
        --env-file=env_file \
        kf1-docker
```

<details>
  <summary>Copy/Paste (CLICK ME)</summary>

```bash
docker run -d --name kf1 -p 0.0.0.0:7707:7707/udp -p 0.0.0.0:7708:7708/udp -p 0.0.0.0:7717:7717/udp -p 0.0.0.0:28852:28852/udp -p 0.0.0.0:28852:28852/tcp -p 0.0.0.0:8075:8075/tcp -p 0.0.0.0:20560:20560/udp -p 0.0.0.0:20560:20560/tcp -v (your server folder):/home/steam/servers/kf -e STEAM_USER=(user) -e STEAM_PASSWORD=(pass) -e STEAM_CODE=(code) --env-file=env_file kf1-docker
```

</details>

## Useful docker commads

- Track installation/server setup after docker run command
  - `docker logs --follow kf1-docker`
- Stop & remove server container
  - `docker stop $(docker ps -aq); docker rm -f Server_name_here`


CHANGELOG.md
# v3.9 ( August 30 - 2021 )

- [feature] Added KF_SPECIMEN_EVENT_TYPE to control ZED skins based on special event types

## v3.8 ( August 30 - 2021 )

- [feature] Added KF_SPECTATORS to modify number of spectators in the server

## v3.7 ( August 22 - 2021 )

- [fix] Wrong paths for steam libs

## v3.6 ( August 21 - 2021 )

- [misc] Updated all scripts with better debugging
- [docker] Updated docker image in docker hub (Increased in size, too)
- [documentation] Added some critical notes regarding OSX users

## v3.5 ( June 13 - 2020 )

- Enhancement PR From @jdbrowndev includes the following changes:
  - Server install is done at run time instead of build time. This keeps the docker image small and allows you to easily tweak the server files post-install using a docker volume.
  - Custom maps and mutators have been removed to provide a vanilla install. These customizations can be added post-install.


scripts/kf1_install.sh
#!/bin/bash

set -o xtrace

if [[ ! -f "/home/steam/servers/kf/System/Core.u" ]]; then
  echo "#### No server install found; preparing to run steamcmd... ####"

  if [[ -z $STEAM_USER ]]; then
    echo "## Env var STEAM_USER is required but not found ##"
    exit 1
  fi

  if [[ -z $STEAM_PASSWORD ]]; then
    echo "## Env var STEAM_PASSWORD is required but not found ##"
    exit 1
  fi

  if [[ -z $STEAM_CODE ]]; then
    echo "## Env var STEAM_CODE not found; this is only a problem if steam guard is enabled ##"
  fi

  /home/steam/steamcmd/steamcmd.sh \
    +login $STEAM_USER $STEAM_PASSWORD $STEAM_CODE \
    +force_install_dir /home/steam/servers/kf \
    +app_update 215360 validate \
    +quit

  if [[ $? -ne 0 ]]; then
    echo "# Install failed, aborting... #"
    exit 1
  fi

  echo "#### Copying run-time functions to /System ####"
  cp main.sh kf1_functions.sh /home/steam/servers/kf/System

  echo "#### Patching with latest Steam Libs to fix VAC/Server List ####"
  cp /home/steam/steamcmd/linux32/libtier0_s.so /home/steam/servers/kf/System/libtier0_s.so
  cp /home/steam/steamcmd/linux32/libvstdlib_s.so /home/steam/servers/kf/System/libvstdlib_s.so
  cp /home/steam/steamcmd/linux32/steamclient.so /home/steam/servers/kf/System/steamclient.so
fi

cd /home/steam/servers/kf/System
./main.sh


scripts/kf1_functions.sh
#!/bin/bash

function require_config() {
  echo "#### Generating INI files ####"
  # Generate INI files for the first time or 'sed' will fail
  if [[ ! -f "KillingFloor.ini" ]]; then
    ./ucc-bin server KF-BioticsLab.rom?game=KFmod.KFGameType?VACSecured=true?MaxPlayers=6 -nohomedir &
    sleep 20
    ps aux
    killall ucc-bin-real
    ps aux
  fi
}

# Mandatory Changes to KillingFloor.ini / Optimisations
function optimize_kf(){
  echo "#### Optimizing Settings ####"
  sed -i "s/ServerBehindNAT=False/ServerBehindNAT=True/g" KillingFloor.ini
  sed -i "s/SendStats=False/SendStats=True/g" KillingFloor.ini
  sed -i "s/AllowDownloads=False/AllowDownloads=True/g" KillingFloor.ini
  sed -i "s/UseCompression=False/UseCompression=True/g" KillingFloor.ini
  sed -i "s/bEnabled=False/bEnabled=True/g" KillingFloor.ini
}

function load_config() {
  ## Load defaults if nothing has been set
  echo "#### Applying Default Options ####"
  ################## MAP SETTINGS ##################
  # Default Launch Map
  [[ -z "$KF_MAP" ]] && export KF_MAP=KF-BioticsLab

  # 1 Beginner, 2 Normal, 4 Hard, 5 Suicidal, 7 HoE
  [[ -z "$KF_DIFFICULTY" ]] && export KF_DIFFICULTY=4

  # Setting this creates a private server
  [[ -z "$KF_GAME_PASS" ]] && export KF_GAME_PASS=""

  # 0 --> 4 waves, 1 --> 7 waves, 2 --> 10 waves
  [[ -z "$KF_GAME_LENGTH" ]] && export KF_GAME_LENGTH=2

  ################## SERVER SETTINGS ##################
  # Name that appears in the server browser
  [[ -z "$KF_SERVER_NAME" ]] && export KF_SERVER_NAME=Killing Floor Server - DOCKERIZED

  # Name that appears in the server browser
  [[ -z "$KF_ADMIN_NAME" ]] && export KF_ADMIN_NAME=Admin

  # Used for web console and in-game logins
  [[ -z "$KF_ADMIN_PASS" ]] && export KF_ADMIN_PASS=123

  # Admin Email
  [[ -z "$KF_ADMIN_EMAIL" ]] && export KF_ADMIN_EMAIL=""
  
  # Spectators Count
  [[ -z "$KF_SPECTATORS" ]] && export KF_SPECTATORS=6

  # Specimens event type/skins
  # Select between the following:
  # ET_None -- Default
  # ET_SummerSideshow -- for the summer ZEDs
  # ET_HillbillyHorror -- for the Halloween ZEDs
  # ET_TwistedChristmas -- for the Christmas ZEDs
  [[ -z "$KF_SPECIMEN_EVENT_TYPE" ]] && export KF_SPECIMEN_EVENT_TYPE=ET_None

  # Server Message Of The Day
  [[ -z "$KF_MOTD" ]] && export KF_MOTD=Welcome To My Server

  # Server redirectURL to download hosted mods from
  [[ -z "$KF_REDIRECT" ]] && export KF_REDIRECT=http://www.skillzservers.com/kf-redirect/

  # Game Port
  [[ -z "$KF_GAME_PORT" ]] && export KF_GAME_PORT=7707

  # Query Port
  [[ -z "$KF_QUERY_PORT" ]] && export KF_QUERY_PORT=7717

  # WebAdmin Port
  [[ -z "$KF_WEBADMIN_PORT" ]] && export KF_WEBADMIN_PORT=8075

  ## Edit KillingFloor.ini
  ############# BASIC REPLACEMENTS #############
  ## sed
  sed -i "s/ServerName=.*/ServerName=$KF_SERVER_NAME/g" KillingFloor.ini
  sed -i "s/AdminName=.*/AdminName=$KF_ADMIN_NAME/g" KillingFloor.ini
  sed -i "s/AdminPassword=.*/AdminPassword=$KF_ADMIN_PASS/g" KillingFloor.ini
  sed -i "s/AdminEmail=.*/AdminEmail=$KF_ADMIN_EMAIL/g" KillingFloor.ini
  sed -i "s/MessageOfTheDay=.*/MessageOfTheDay=$KF_MOTD/g" KillingFloor.ini
  sed -i "s,RedirectToURL=.*,RedirectToURL=$KF_REDIRECT,g" KillingFloor.ini
  sed -i "s/MaxSpectators=.*/MaxSpectators=$KF_SPECTATORS/g" KillingFloor.ini
  sed -i "s/SpecialEventType=.*/SpecialEventType=$KF_SPECIMEN_EVENT_TYPE/g" KillingFloor.ini
  sed -i "s/Port=7707/Port=$KF_GAME_PORT/g" KillingFloor.ini
  sed -i "s/OldQueryPortNumber=.*/OldQueryPortNumber=$KF_QUERY_PORT/g" KillingFloor.ini
  sed -i "s/ListenPort=.*/ListenPort=$KF_WEBADMIN_PORT/g" KillingFloor.ini
  ## grep
  grep -q "KFGameLength=" KillingFloor.ini && sed -i "s/KFGameLength=.*/KFGameLength=$KF_GAME_LENGTH/g" KillingFloor.ini || echo -e "[KFMod.KFGameType]\nKFGameLength=$KF_GAME_LENGTH" >> KillingFloor.ini
  grep -q "GameDifficulty=" KillingFloor.ini && sed -i "s/GameDifficulty=.*/GameDifficulty=$KF_DIFFICULTY/g" KillingFloor.ini || sed -i "/AccessControlClass=Engine.AccessControl/a GameDifficulty=$KF_DIFFICULTY" KillingFloor.ini
  ### grep optimizations
  grep -q "bVACSecured=True" KillingFloor.ini || sed -i "/AccessControlClass=Engine.AccessControl/a bVACSecured=True" KillingFloor.ini
  grep -q "bAdminCanPause=True" KillingFloor.ini || sed -i "/AccessControlClass=Engine.AccessControl/a bAdminCanPause=True" KillingFloor.ini
  grep -q "bAllowBehindView=True" KillingFloor.ini || sed -i "/AccessControlClass=Engine.AccessControl/a bAllowBehindView=True" KillingFloor.ini
  grep -q "bWeaponShouldViewShake=False" KillingFloor.ini || sed -i "/AccessControlClass=Engine.AccessControl/a bWeaponShouldViewShake=False" KillingFloor.ini
  ## conditional
  [[ -z "$KF_GAME_PASS" ]] || sed -i "s/GamePassword=.*/GamePassword=$KF_GAME_PASS/g" KillingFloor.ini
}

function launch() {
  echo "#### Launching Server ####"
  local cmd

  cmd="./ucc-bin server "
  cmd+="$KF_MAP.rom?game=KFmod.KFGameType?"
  [[ -z "$KF_MUTATORS" ]] || cmd+="mutator=$KF_MUTATORS?"
  cmd+="VACSecured=true?MaxPlayers=6 -nohomedir"
  echo "## Server launched with command: $cmd ##" > $0-cmd.log
  exec $cmd
}
